deltacat/compute/compactor_v2/utils/io.py

```diff
@@ -0,0 +1,178 @@
+    result = []
+
+    delta_source_incremental_deltas = io_v1._discover_deltas(
+        delta_source_partition_locator,
+        previous_compacted_high_watermark,
+        last_stream_position_to_compact,
+        deltacat_storage,
+        deltacat_storage_kwargs,
+        list_deltas_kwargs,
+    )
+
+    result.extend(delta_source_incremental_deltas)
+
+    logger.info(
+        f"Length of input deltas from delta source table is {len(delta_source_incremental_deltas)}"
+        f" from ({previous_compacted_high_watermark}, {last_stream_position_to_compact}]"
+    )
+
+    if rebase_source_partition_locator:
+        previous_compacted_deltas = io_v1._discover_deltas(
+            source_partition_locator,
+            None,
+            None,
+            deltacat_storage,
+            deltacat_storage_kwargs,
+            list_deltas_kwargs,
+        )
+
+        result.extend(previous_compacted_deltas)
+
+        logger.info(
+            f"Length of input deltas from previous compacted table is {len(previous_compacted_deltas)}"
+            f" from ({None}, {None}]"
+        )
+
+    return result
+
+
+def create_uniform_input_deltas(
+    input_deltas: List[Delta],
+    hash_bucket_count: int,
+    compaction_audit: CompactionSessionAuditInfo,
+    min_delta_bytes: Optional[float] = MIN_DELTA_BYTES_IN_BATCH,
+    min_file_counts: Optional[float] = MIN_FILES_IN_BATCH,
+    previous_inflation: Optional[float] = PYARROW_INFLATION_MULTIPLIER,
+    deltacat_storage=unimplemented_deltacat_storage,
+    deltacat_storage_kwargs: Optional[Dict[str, Any]] = {},
+) -> List[DeltaAnnotated]:
+
+    delta_bytes = 0
+    delta_manifest_entries = 0
+    estimated_da_bytes = 0
+    input_da_list = []
+
+    for delta in input_deltas:
+        manifest_entries = delta.manifest.entries
+        delta_manifest_entries += len(manifest_entries)
+        for entry_index in range(len(manifest_entries)):
+            entry = append_content_type_params(
+                delta=delta,
+                entry_index=entry_index,
+                deltacat_storage=deltacat_storage,
+                deltacat_storage_kwargs=deltacat_storage_kwargs,
+            )
+
+            delta_bytes += entry.meta.content_length
+            estimated_da_bytes += estimate_manifest_entry_size_bytes(
+                entry=entry, previous_inflation=previous_inflation
+            )
+
+        delta_annotated = DeltaAnnotated.of(delta)
+        input_da_list.append(delta_annotated)
+
+    logger.info(f"Input deltas to compact this round: " f"{len(input_da_list)}")
+    logger.info(f"Input delta bytes to compact: {delta_bytes}")
+    logger.info(f"Input delta files to compact: {delta_manifest_entries}")
+
+    if not input_da_list:
+        raise RuntimeError("No input deltas to compact!")
+
+    size_estimation_function = functools.partial(
+        func=estimate_manifest_entry_size_bytes, previous_inflation=previous_inflation
+    )
+
+    rebatched_da_list = DeltaAnnotated.rebatch(
```

Comment:
So in v2, our default will split the annotated deltas into row group first then do an original rebatch? So in the case of small files, the `row_groups_to_download` will always be 1 so that we will not split on a small file?

Reply:
Yeah, the lowest split granularity we can go is at the row group level.